File: lib/rich-text.ts
interface RawInlineContent {
  text?: string;
}

interface RawBlock {
  content?: RawInlineContent[] | string;
  children?: RawBlock[];
}

/**
 * BlogPostCard's list-view excerpt needs plain prose, not BlockNote's own
 * Block[] JSON (RichTextEditor.tsx) that `description` is actually stored
 * as -- slicing the raw JSON string itself (the pre-editor behavior) would
 * show a garbled blob of `{"type":"paragraph"...}` on the Blog list page.
 * Deliberately dependency-free (no `@blocknote/core` import) so this stays
 * safe to call from a plain Server Component without pulling the editor's
 * client bundle along -- it only needs to read `content`/`children`, the
 * same two keys every Block shape (including this app's own custom
 * `layoutImage`) already carries.
 */
export function extractPlainText(raw: string | null | undefined): string {
  if (!raw) return '';
  let blocks: RawBlock[];
  try {
    const parsed: unknown = JSON.parse(raw);
    if (!Array.isArray(parsed)) return raw;
    blocks = parsed;
  } catch {
    // Not JSON -- a legacy plain-text post (predates this editor).
    return raw;
  }

  const parts: string[] = [];
  const walk = (list: RawBlock[]) => {
    for (const block of list) {
      if (typeof block.content === 'string') {
        parts.push(block.content);
      } else if (Array.isArray(block.content)) {
        for (const inline of block.content) {
          if (typeof inline.text === 'string') parts.push(inline.text);
        }
      }
      if (block.children?.length) walk(block.children);
    }
  };
  walk(blocks);

  return parts.join(' ').replace(/\s+/g, ' ').trim();
}

File: android/app/src/main/java/com/trails/app/ui/blog/RichText.kt
package com.trails.app.ui.blog

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * lib/rich-text.ts::extractPlainText, ported 1:1 -- Blog Post `description`
 * is stored as BlockNote's own Block[] JSON (see the Android app's Phase 4
 * plan for offline drafting); this app only ever reads it, walking the same
 * `content`/`children` keys every block shape carries, dependency-free (no
 * BlockNote-equivalent rendering here at all -- a real rich-text renderer is
 * future work, this is the same plain-prose fallback the web's own list-view
 * excerpt uses).
 */
fun extractPlainText(raw: String?): String {
    if (raw.isNullOrBlank()) return ""
    val parsed = runCatching { Json.parseToJsonElement(raw) }.getOrNull() ?: return raw
    val blocks = parsed as? JsonArray ?: return raw

    val parts = mutableListOf<String>()
    fun walk(list: JsonArray) {
        for (blockEl in list) {
            val block = blockEl as? JsonObject ?: continue
            when (val content = block["content"]) {
                is JsonPrimitive -> if (content.isString) parts.add(content.content)
                is JsonArray -> for (inline in content) {
                    val text = (inline as? JsonObject)?.get("text") as? JsonPrimitive
                    if (text != null && text.isString) parts.add(text.content)
                }
                else -> Unit
            }
            (block["children"] as? JsonArray)?.let { if (it.isNotEmpty()) walk(it) }
        }
    }
    walk(blocks)

    return parts.joinToString(" ").replace(Regex("\\s+"), " ").trim()
}
